### ISSUE TITLE
Control of the legend text in structure data chart

### ISSUE THREAD

**Issue by jboffel** [goat1000/SVGGraph#41]
Hi,

I'd like to be able to control the legend to actually display colors and text aggregated per the selected column.

Example:
|Name|sex|
|Bob   |  M |
|Bob   |  M |
|Bob   |  M |
|Bob   |  M |

**Comment by goat1000** [comment]
I'll need to have a good think about this. Maybe adding a callback function that allows you to change the list of entries after it has been created would work.

In the meantime, SVGGraph will only display entries where they exist, so you can get it working by adding an extra field to one each of the male and female rows:

```PHP
$settings = [
  'structure' => [
    'key' => 0, 'value' => 3,
    'axis_text' => 1,
    'colour' => 4,
    'legend_text' => 5,
  ],
  'show_legend' => true,
];

$m = '#33f';
$f = '#f33';

$values = [
  [0, 'Bob', 'M', 20, $m, 'Male'],
  [1, 'Ena', 'F', 30, $f, 'Female'],
  [2, 'Jon', 'M', 18, $m],
  [3, 'Bob', 'M', 60, $m],
];

$graph = new Goat1000\SVGGraph\SVGGraph(600,400,$settings);
$graph->values($values);
$graph->render('BarGraph');
```

**Comment by jboffel** [comment]
Thanks for the help unfortunately this approach does not work reliably, I already tried it.
Example:
$values = [
  [0, 'Bob', 'M', 20, $m, 'Male'],
  [1, 'Jon', 'M', 18, $m],
  [2, 'Ena', 'F', 30, $f, 'Female'],
  [3, 'Bob', 'M', 60, $m],
];

This one will only show 1 legend entry for 'Male'. Seems if there is a value without the legend it stop to look forward for more entries;

**Comment by goat1000** [comment]
That sounds like a bug - though it appears to be working here. Which versions of SVGGraph and PHP are you using?

**Comment by jboffel** [comment]
Hi,

"goat1000/svggraph": "^3.3"
PHP 7.1.16 (cli)

**Comment by goat1000** [comment]
I haven't got a PHP 7.1 to test with, but it looks like it is working properly in versions 5.4 and 7.3.

Do you have a script I can look at where it consistently fails?

Thanks.

**Comment by jboffel** [comment]
Yes I have but I can't just post the data, I would need to obfuscate it.
Instead I went on debugging this case. It's related to the goat1000/svggraph/Legend.php file.

Line 89 
`if(!$this->show_empty && !$this->graph->isVisible($item, $dataset))`

My data has a big number of entries, over 1000 and the design I choose make the bar to overlap each other and might easily hide some of them and I guess the isVisible function in that case prevent the legend from being added to the entry list. If it end of one of dataset value where I put the legend text then it will not show up...

**Comment by goat1000** [comment]
The `isVisible` function is very simple, for bar graphs it just checks if the value is not 0 or `null`. It doesn't check if bars are overlapping or anything like that. 

Try setting the `legend_show_empty` option to `true` and see if that makes it work properly. If it does, it means the bar that was assigned legend entry text has a value of 0 or `null`.

**Comment by jboffel** [comment]
Actually you have a point as to make it work I also had to add put show_data_labels to true, but would have been the same with your proposal:

```
  protected function setBarLegendEntry($dataset, $index, DataItem $item)
  {
    if($this->legend_show_empty || $this->show_data_labels || $item->value != 0) {
      $bar = ['fill' => $this->getColour($item, $index, $dataset)];
      $this->setStroke($bar, $item, $dataset);
      $this->setLegendEntry($dataset, $index, $item, $bar);
    }
  }
```

But actually, I'm using floating bar graph with a range of value by giving 2 columns from the data set values. And of course data at the beginning of the array might start from 0 to N as a range, but that should not prevent the legend from showing up I believe.

**Comment by goat1000** [comment]
Ah, that's the problem - the bug is that floating bar graphs should override the `isVisible` function. That should be easy enough to fix.

Thanks for helping me figure this out.

**Comment by goat1000** [comment]
I've uploaded the fix for the floating bar legend entries not showing up in commit e4289d4 - I've simplified how it works in some of the other graph types too.

I'll see if I can come up with a better way of handling legend entries in the new year.
